finality-aleph/src/network.rs

```diff
@@ -196,22 +209,29 @@ impl Peers {
                 self.to_peer.entry(*session_id).and_modify(|hm| {
                     hm.remove(node_id);
                 });
+                self.session_peers
+                    .entry(*session_id)
+                    .and_modify(|v| v.retain(|p| *p != *peer));
             }
         }
-        self.to_peer.retain(|_, hm| !hm.is_empty());
+        self.session_peers.retain(|_, v| !v.is_empty());
     }
 
     fn peers_authenticated_for(&self, session_id: SessionId) -> impl Iterator<Item = &PeerId> {
-        self.to_peer
-            .get(&session_id)
-            .into_iter()
-            .map(|hm| hm.values())
-            .flatten()
+        self.session_peers.get(&session_id).into_iter().flatten()
     }
 
     fn get(&self, session_id: SessionId, node_id: NodeIndex) -> Option<&PeerId> {
         self.to_peer.get(&session_id)?.get(&node_id)
     }
+
+    fn get_rand(&self, session_id: SessionId, n_peers: usize) -> impl Iterator<Item = &PeerId> {
+        self.session_peers
+            .get(&session_id)
+            .expect("the session already started")
+            .as_slice()
+            .choose_multiple(&mut rand::thread_rng(), n_peers)
```

Comment:
Will it work properly if the current number of peers is `< n_peers`? In which case it should choose all.

Reply:
Interestingly the `RandIterator` version of the `choose_multiple` method states directly that this is the case (returns all if  `n_peers` is too big), but this is not specified in this variant...

Reply:
Worked in previous version and works in the new one.